Userland/Libraries/LibWeb/Communication/WebSocket.h

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (c) 2021, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/ByteBuffer.h>
+#include <AK/RefCounted.h>
+#include <AK/URL.h>
+#include <AK/Weakable.h>
+#include <LibCore/Object.h>
+#include <LibWeb/Bindings/WindowObject.h>
+#include <LibWeb/Bindings/Wrappable.h>
+#include <LibWeb/DOM/EventTarget.h>
+#include <LibWeb/DOM/ExceptionOr.h>
+#include <LibWeb/Forward.h>
+
+#define ENUMERATE_WEBSOCKET_EVENT_HANDLERS(E)  \
+    E(onerror, HTML::EventNames::error)        \
+    E(onclose, HTML::EventNames::close)        \
+    E(onopen, Communication::EventNames::open) \
+    E(onmessage, Communication::EventNames::message)
+
+namespace Protocol {
+class WebSocketClient;
+class WebSocket;
+}
+
+namespace Web::Communication {
+
+class WebSocketHandler : public Core::Object {
+    C_OBJECT(WebSocketHandler)
+public:
+    static WebSocketHandler& the();
+
+    RefPtr<Protocol::WebSocket> connect(const URL&);
+
+private:
+    WebSocketHandler();
+    RefPtr<Protocol::WebSocketClient> m_websocket_client;
+};
+
+class WebSocket final
+    : public RefCounted<WebSocket>
+    , public Weakable<WebSocket>
+    , public DOM::EventTarget
+    , public Bindings::Wrappable {
+public:
+    enum class ReadyState : u16 {
+        Connecting = 0,
+        Open = 1,
+        Closing = 2,
+        Closed = 3,
+    };
+
+    using WrapperType = Bindings::WebSocketWrapper;
+
+    static NonnullRefPtr<WebSocket> create(DOM::Window& window, URL& url)
+    {
+        return adopt_ref(*new WebSocket(window, url));
+    }
+
+    static DOM::ExceptionOr<NonnullRefPtr<WebSocket>> create_with_global_object(Bindings::WindowObject& window, const String& url);
+
+    virtual ~WebSocket() override;
+
+    using RefCounted::ref;
+    using RefCounted::unref;
+
+    String url() const { return m_url.to_string(); }
+
+#undef __ENUMERATE
+#define __ENUMERATE(attribute_name, event_name)    \
+    void set_##attribute_name(HTML::EventHandler); \
+    HTML::EventHandler attribute_name();
+    ENUMERATE_WEBSOCKET_EVENT_HANDLERS(__ENUMERATE)
+#undef __ENUMERATE
+
+    void set_event_handler_attribute(const FlyString& name, HTML::EventHandler);
+    HTML::EventHandler get_event_handler_attribute(const FlyString& name);
+
+    ReadyState ready_state() const;
+    String extensions() const;
+    String protocol() const;
+
+    String binary_type() { return m_binary_type; };
```

Comment:
hoping that the wrapper generator can deal with this:

```suggestion
    const String& binary_type() { return m_binary_type; };
```